File: app/src/main/java/za/co/gundula/app/arereyeng/ui/BaseActivity.java
package za.co.gundula.app.arereyeng.ui;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.PersistableBundle;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;

import com.google.android.gms.auth.api.Auth;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.common.api.ResultCallback;
import com.google.android.gms.common.api.Status;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

import za.co.gundula.app.arereyeng.Constants;
import za.co.gundula.app.arereyeng.R;

/**
 * Created by kgundula on 2016/11/07.
 */

public abstract class BaseActivity extends AppCompatActivity implements
        GoogleApiClient.OnConnectionFailedListener, GoogleApiClient.ConnectionCallbacks {

    public static String LOG_TAG = BaseActivity.class.getName();
    protected String mProvider, mEncodedEmail;

    /* Client used to interact with Google APIs. */
    protected GoogleApiClient mGoogleApiClient;
    protected FirebaseAuth mAuth;
    protected FirebaseAuth.AuthStateListener mAuthListener;


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        /* Setup the Google API object to allow Google logins */
        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.default_web_client_id))
                .requestEmail()
                .build();

        /**
         * Build a GoogleApiClient with access to the Google Sign-In API and the
         * options specified by gso.
         */

        /* Setup the Google API object to allow Google+ logins */
        mGoogleApiClient = new GoogleApiClient.Builder(this)
                .enableAutoManage(this /* FragmentActivity */, this /* OnConnectionFailedListener */)
                .addApi(Auth.GOOGLE_SIGN_IN_API, gso)
                .build();


        final SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(BaseActivity.this);
        /* Get mEncodedEmail and mProvider from SharedPreferences, use null as default value */
        mEncodedEmail = sp.getString(Constants.key_encoded_email, null);
        mProvider = sp.getString(Constants.key_provider, null);

        if (!((this instanceof LoginActivity) || (this instanceof RegisterActivity))) {


            mAuth = FirebaseAuth.getInstance();
            mAuthListener = new FirebaseAuth.AuthStateListener() {
                @Override
                public void onAuthStateChanged(@NonNull FirebaseAuth firebaseAuth) {
                    FirebaseUser user = firebaseAuth.getCurrentUser();
                    if (user != null) {
                        Log.i("Ygritte", "User UID : " + user.getUid());
                    } else {
                        SharedPreferences.Editor spe = sp.edit();
                        spe.putString(Constants.key_encoded_email, "");
                        spe.putString(Constants.key_provider, "");
                        spe.apply();
                        takeUserToLoginScreenOnUnAuth();
                    }

                }
            };

            mAuth.addAuthStateListener(mAuthListener);
        }

    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        /* Cleanup the AuthStateListener */
        if (!((this instanceof LoginActivity) || (this instanceof RegisterActivity))) {
            mAuth.removeAuthStateListener(mAuthListener);
        }

    }

    @Override
    public void onSaveInstanceState(Bundle outState, PersistableBundle outPersistentState) {
        super.onSaveInstanceState(outState, outPersistentState);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        if (id == android.R.id.home) {
            super.onBackPressed();
            return true;
        }

        if (id == R.id.action_logout) {
            logout();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }


    private void takeUserToLoginScreenOnUnAuth() {
        /* Move user to Login Activity, and remove the BackStack */
        Intent intent = new Intent(BaseActivity.this, LoginActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        startActivity(intent);
        finish();
    }

    /**
     * Logs out the user from their current session and starts LoginActivity.
     * Also disconnects the mGoogleApiClient if connected and provider is Google
     */
    protected void logout() {
        /* Logout if mProvider is not null */
        if (mProvider != null) {
            FirebaseAuth.getInstance().signOut();
            if (mProvider.equals(Constants.google_provider)) {
                /* Logout from Google+ */
                Auth.GoogleSignInApi.signOut(mGoogleApiClient).setResultCallback(
                        new ResultCallback<Status>() {
                            @Override
                            public void onResult(Status status) {
                                //nothing
                            }
                        });
            }
        }
    }

    @Override
    public void onConnectionFailed(ConnectionResult connectionResult) {

    }

}

File: app/src/main/java/za/co/gundula/app/arereyeng/ui/BaseActivity.kt
package za.co.gundula.app.arereyeng.ui

import android.content.Intent
import android.os.Bundle
import android.os.PersistableBundle
import android.preference.PreferenceManager
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import com.google.android.gms.auth.api.Auth
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.common.api.ResultCallback
import com.google.android.gms.common.api.Status
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseAuth.AuthStateListener
import za.co.gundula.app.arereyeng.R
import za.co.gundula.app.arereyeng.ui.BaseActivity
import za.co.gundula.app.arereyeng.utils.Constants

/**
 * Created by kgundula on 2016/11/07.
 */
abstract class BaseActivity : AppCompatActivity(), GoogleApiClient.OnConnectionFailedListener, GoogleApiClient.ConnectionCallbacks {
    protected var mProvider: String? = null
    protected var mEncodedEmail: String? = null

    /* Client used to interact with Google APIs. */
    @JvmField
    protected var mGoogleApiClient: GoogleApiClient? = null
    protected var mAuth: FirebaseAuth? = null
    protected var mAuthListener: AuthStateListener? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        /* Setup the Google API object to allow Google logins */
        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.default_web_client_id))
                .requestEmail()
                .build()
        /**
         * Build a GoogleApiClient with access to the Google Sign-In API and the
         * options specified by gso.
         */

        /* Setup the Google API object to allow Google+ logins */mGoogleApiClient = GoogleApiClient.Builder(this)
                .enableAutoManage(this /* FragmentActivity */, this /* OnConnectionFailedListener */)
                .addApi(Auth.GOOGLE_SIGN_IN_API, gso)
                .build()
        val sp = PreferenceManager.getDefaultSharedPreferences(this@BaseActivity)
        /* Get mEncodedEmail and mProvider from SharedPreferences, use null as default value */mEncodedEmail = sp.getString(Constants.key_encoded_email, null)
        mProvider = sp.getString(Constants.key_provider, null)
        if (!(this is LoginActivity || this is RegisterActivity)) {
            mAuth = FirebaseAuth.getInstance()
            mAuthListener = AuthStateListener { firebaseAuth ->
                val user = firebaseAuth.currentUser
                if (user != null) {
                    //Log.i("Ygritte", "User UID : " + user.getUid());
                } else {
                    val spe = sp.edit()
                    spe.putString(Constants.key_encoded_email, "")
                    spe.putString(Constants.key_provider, "")
                    spe.apply()
                    takeUserToLoginScreenOnUnAuth()
                }
            }
            mAuth!!.addAuthStateListener(mAuthListener)
        }
    }

    public override fun onDestroy() {
        super.onDestroy()
        /* Cleanup the AuthStateListener */if (!(this is LoginActivity || this is RegisterActivity)) {
            mAuth!!.removeAuthStateListener(mAuthListener!!)
        }
    }

    override fun onSaveInstanceState(outState: Bundle, outPersistentState: PersistableBundle) {
        super.onSaveInstanceState(outState, outPersistentState)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == android.R.id.home) {
            super.onBackPressed()
            return true
        }
        if (id == R.id.action_logout) {
            logout()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun takeUserToLoginScreenOnUnAuth() {
        /* Move user to Login Activity, and remove the BackStack */
        val intent = Intent(this@BaseActivity, LoginActivity::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        startActivity(intent)
        finish()
    }

    /**
     * Logs out the user from their current session and starts LoginActivity.
     * Also disconnects the mGoogleApiClient if connected and provider is Google
     */
    protected fun logout() {
        /* Logout if mProvider is not null */
        if (mProvider != null) {
            FirebaseAuth.getInstance().signOut()
            if (mProvider == Constants.google_provider) {
                /* Logout from Google+ */
                Auth.GoogleSignInApi.signOut(mGoogleApiClient).setResultCallback(
                        object : ResultCallback<Status?> {
                            override fun onResult(status: Status?) {
                                //nothing
                            }
                        })
            }
        }
    }

    override fun onConnectionFailed(connectionResult: ConnectionResult) {}

    companion object {
        var LOG_TAG = BaseActivity::class.java.name
    }
}